lib/hammer_cli_foreman/api/oauth/password_grant.rb

```diff
@@ -0,0 +1,81 @@
+require 'hammer_cli_foreman/openid_connect'
+
+module HammerCLIForeman
+  module Api
+    module Oauth
+      class PasswordGrant < ApipieBindings::Authenticators::TokenAuth
+        attr_accessor :oidc_token_endpoint, :oidc_client_id, :user, :password, :token
+
+        def initialize(oidc_token_endpoint, oidc_client_id, user, password)
+          @oidc_token_endpoint = oidc_token_endpoint
+          @oidc_client_id = oidc_client_id
+          @user = user
+          @password = password
+          super set_token(oidc_token_endpoint, oidc_client_id, user, password)
+        end
+
+        def authenticate(request, token)
+          if HammerCLI.interactive?
+            set_token_interactively
+          end
+          super
+        end
+
+        def set_token_interactively
+          @token ||= set_token(get_oidc_token_endpoint, get_oidc_client_id, get_user, get_password)
+        end
+
+        def set_token(input_oidc_token_endpoint, input_oidc_client_id, input_user, input_password)
+          @oidc_token_endpoint = input_oidc_token_endpoint if input_oidc_token_endpoint
+          @user = input_user
+          @password = input_password
+          @oidc_client_id = input_oidc_client_id if input_oidc_client_id
+          if @user && @password && @oidc_token_endpoint && @oidc_client_id
+            @token = HammerCLIForeman::OpenidConnect.new(
+              @oidc_token_endpoint, @oidc_client_id).get_token(@user, @password)
+          else
+            @token = nil
+          end
+        end
+
+        def error(ex)
+          if ex.is_a?(RestClient::InternalServerError)
+            self.clear
```

Comment:
When logging in with invalid password I get 
[ERROR 2019-10-25T15:37:22 Exception] Error: undefined method `clear' for #<HammerCLIForeman::Api::Oauth::PasswordGrant:0x0000000002fb0098>
Error: undefined method `clear' for #<HammerCLIForeman::Api::Oauth::PasswordGrant:0x0000000002fb0098>

Reply:
The clear should probably remove the values from the object

Reply:
And that's why we need two reviewers... When I looked at the code I didn't notice that there is no method `clear` for `AuthenticationCodeGrant` and `PasswordGrant` objects.

@rahulbajaj0509, is it a leftover or you forget to define those methods for `PasswordGrant` and `AuthenticationCodeGrant` objects? This method exists only in `InteractiveBasicAuth`.